core/src/utils/prove_distributed/mod.rs

```diff
@@ -0,0 +1,241 @@
+    utils::{
+        baby_bear_poseidon2::{Perm, Val},
+        BabyBearPoseidon2, SP1CoreOpts, SP1CoreProverError,
+    },
+};
+
+mod checkpoints;
+mod threads;
+
+pub type CoreSC = BabyBearPoseidon2;
+
+pub type Checkpoint = ExecutionState;
+pub type Shard = ExecutionRecord;
+pub type ShardsPublicValues = Vec<Val>;
+pub type Commitments = Vec<Commitment>;
+pub type Commitment = Hash<Val, Val, 8>;
+pub type PartialProofs = Vec<ShardProof<BabyBearPoseidon2>>;
+pub type Challenger = DuplexChallenger<Val, Perm, 16, 8>;
+pub type Machine = StarkMachine<BabyBearPoseidon2, RiscvAir<Val>>;
+pub type ProvingKey = StarkProvingKey<BabyBearPoseidon2>;
+pub type VerifyingKey = StarkVerifyingKey<BabyBearPoseidon2>;
+
+// This is the entry point of the orchestrator
+// It computes the checkpoints to be sent to the workers
+pub fn compute_checkpoints(
+    stdin: &SP1Stdin,
+    program: &Program,
+    nb_workers: usize,
+    context: SP1Context,
+) -> Result<
+    (
+        Vec<(Checkpoint, usize)>,
+        Checkpoint, // last checlpoint
+        Vec<u8>,
+        PublicValues<u32, u32>,
+        SP1CoreOpts,
+        u32, // total nb shards
+    ),
+    SP1CoreProverError,
+> {
+    log::info!("Computing checkpoints");
+    let opts = SP1CoreOpts::default();
+
+    // Execute the program.
+    let mut runtime = Runtime::with_context(program.clone(), opts, context);
+    runtime.write_vecs(&stdin.buffer);
+
+    for proof in stdin.proofs.iter() {
+        runtime.write_proof(proof.0.clone(), proof.1.clone());
+    }
+
+    let mut checkpoints_states = Vec::new();
+
+    // Execute the program, saving checkpoints at the start of every `shard_batch_size` cycle range.
+    let create_checkpoints_span = tracing::debug_span!("create checkpoints").entered();
+
+    let (total_nb_shards, public_values_stream, public_values) = loop {
+        // Execute the runtime until we reach a checkpoint.
+        let (checkpoint, done) = runtime
+            .execute_state()
+            .map_err(SP1CoreProverError::ExecutionError)?;
+
+        let mut checkpoint_file = tempfile::tempfile().map_err(SP1CoreProverError::IoError)?;
+
+        checkpoint
+            .save(&mut checkpoint_file)
+            .map_err(SP1CoreProverError::IoError)?;
+
+        checkpoints_states.push(checkpoint_file);
+
+        // If we've reached the final checkpoint, break out of the loop.
+        if done {
+            break (
+                runtime.shard() as usize,
+                runtime.state.public_values_stream,
+                runtime
+                    .records
+                    .last()
+                    .expect("at least one record")
+                    .public_values,
+            );
+        }
+    };
+
+    create_checkpoints_span.exit();
+
+    log::info!(
+        "Nb shards: {}",
+        checkpoints_states.len() * opts.shard_batch_size
+    );
+
+    let nb_checkpoints_per_workers =
+        (checkpoints_states.len() as f64 / nb_workers as f64).ceil() as usize;
```

Comment:
This can lead to work imbalance

For example dividing 40 items on 12 workers will lead to
a base_chunk_size of 40/12 = 3 so work on the first 11 workers
will be 3 * 11 = 33, and the remainder 7 on the last worker.

Instead of dividing 40 work items on 12 cores into:
3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 7 = 3*11 + 7 = 40
the best scheme will divide into
4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3 = 4*4 + 3*8 = 40

See https://github.com/mratsim/constantine/blob/master/constantine/threadpool/partitioners.nim#L26-L77

Reply:
Good catch ! I will study those formulas, but is it worth it to implement such a mechanism ? 

Here i put a `ceil()` call that maximize the number of shards on each workers. While it is clearly not ideal, only the last(s) will have a reduced load, which is not a problem since equal load means equal processing time then all the workers will finish at about the same time, and so it won't add any significant time to the proving process.